[LBP] 김준 3, 4단계 과제 제출합니다(피드백 반영)

kjoon418 opened the pull request:
### 스터디 정보
 - 스터디명: LBP
 - 리뷰이: 김준
 - 리뷰어: @YehyeokBang

### 기존 코드 리팩터링
 - 피드백을 반영해 기존 사칙연산 로직을 수정했습니다.
   - 각 연산자가 스스로 연산에 대한 책임을 가지도록 역할을 분리했습니다.
     책임이 분리되어 코드의 가독성이 증가하고, 기능 추가시에도 기존 코드가 영향받지 않도록 개선했습니다.
   - 오버플로우와 언더플로우를 한 번에 검증하도록 검증 메서드를 수정했습니다.
     (`long` 값을 받아, 해당 값이 `int` 범위 밖에 있는지를 확인하도록 수정했습니다.)
 - 사칙연산 메서드의 파라미터명을 수정했습니다.
   - 어떤 역할을 하는 파라미터인지 명확히 설명할 수 있도록, 좌측 피연산자(`leftOperand`)와 우측 피연산자(`rightOperand`)로 수정했습니다.
 - 테스트 로직이 Assertj의 메서드를 사용하도록 수정했습니다.

### 질문
 - 깃허브의 `Files changed`가 어지럽게 섞여 있습니다.
   - 기존 코드 삭제와 새로운 코드 추가가 섞여서 이런 것 같은데, 이런 경우에 깃허브의 `Files changed`에 깔끔하게 표기되도록 만들 수 있나요?
 - 피드백해 주신 대로 조건문을 메서드로 추출하여 의미를 부여했습니다.
   - 가독성은 좋아진 대신 클래스에 메서드가 너무 많아졌는데, 이 부분은 크게 문제되지 않나요?
 - 동작을 설명하기 위해 Enum의 추상 메서드 이름을 `operate`, `convertListIntoString`으로 지었습니다.
   - 함수형 인터페이스 관례에 따라 `apply`로 짓는 것이 더 나을까요?
 - 3단계의 요구사항을 수행하는 메서드를 만드는데, 메서드 이름만으로 동작을 직관적으로 설명하기 어려워 주석을 통해 문서화했습니다.
   - 요구사항에서 `RuntimeException`을 `throw`하라 했기에 모든 에외 상황에 `RuntimeException`을 던지도록 했습니다.
     그에 따라 문서화 과정에서 `@throws RuntimeException` 부분에 모든 예외 상황을 서술해야 했습니다.
     좋은 문서화가 아닌 것 같은데, 어떤 방법으로 개선할 수 있을까요?
    - 문서화에 미숙한 부분이 있다면 조언 부탁드립니다!
  - 커스텀 구분자의 시작점(`//`)과 끝점(`\n`)을 상수로 표현했습니다.
    그러나 해당 상수가 `Calculator`와 `CalculatorTest` 모두에 필요한 상황인데 어떻게 처리하는 것이 바람직할까요?
    1. 각 클래스에 각각 선언한다(코드 중복 발생).
    2. `Calculator` 클래스의 상수를 `public`으로 열어 둔다.
    3. 상수 값을 `public static`으로 보관할 클래스를 따로 만든다.
    4. (스프링 프로젝트일 경우에) `application.yml`에 값을 입력하고, `@Value` 어노태이션으로 받아오도록 한다.
 - 문자열을 나눌 때 기존에 자주 사용하던 `StringTokenizer`를 사용했습니다.
   - 속도 측면에서 `StringTokenizer`가  `String.split()`보다 좋다고 알고 있는데, 구현 힌트에는 `String.split()`을 사용하라고 적혀 있는 이유가 있나요?

질문이 많음에도 상세히 답변해 주셔서 정말 감사합니다!

kjoon418 committed:
build: Assertj에 대한 의존성 추가

kjoon418 committed:
feat: 3단계 과제 수행

문자열 계산기 기능을 추가함.

kjoon418 committed:
feat: 4단계 과제 수행

3단계에 대한 테스트 케이스를 추가함.
Assertj를 사용하도록 테스트 로직을 리팩토링함.

kjoon418 committed:
refactor: 기존 사칙연산 로직 리팩터링

kjoon418 committed:
chore: 오류 메시지 수정

kjoon418 committed:
fix: 불필요한 변수 제거

kjoon418 commented on src/test/java/CalculatorTest.java:
```diff
@@ -397,7 +396,187 @@ void negative_negative() {
+            assertThat(actualResult).isEqualTo(expectedResult);
         }
+
+        @Test
+        @DisplayName("한 글자인 커스텀 구분자를 사용한다")
+        void customSeparator_oneLetter() {
+            // given
+            String input1 = Separator.CUSTOM.convertListIntoString(TEST_NUMBERS, "^");
+            String input2 = Separator.CUSTOM.convertListIntoString(TEST_NUMBERS, "$");
+            int expectedResult = getSumOfList(TEST_NUMBERS);
+
+            // when
+            int actualResult1 = calculator.addAll(input1);
+            int actualResult2 = calculator.addAll(input2);
+
+            // then
+            assertThat(actualResult1).isEqualTo(expectedResult);
+            assertThat(actualResult2).isEqualTo(expectedResult);
+        }
```
위와 같이 한 테스트 메서드 안에서 여러 케이스를 검사해도 괜찮을까요?
테스트의 주 목적은 **한 글자인 커스텀 구분자 사용**이기에, 위의 테스트를 `"커스텀 구분자로 ^를 사용한다"`와 `"커스텀 구분자로 $를 사용한다"`로 나누는 것이 바람직하지 않다 생각했습니다.

YehyeokBang replied:
`@ParameterizedTest`에 대해 학습해 보시면 해결에 도움이 될 것 같아요. 😄

kjoon418 replied:
이런 좋은 어노태이션을 모르고 있었군요. 잘 활용해서 테스트 코드를 수정하도록 하겠습니다.
피드백 감사합니다!

kjoon418 committed:
fix: 잘못된 공백 제거

kjoon418 committed:
chore: 유틸 메서드들을 모두 static으로 변경

YehyeokBang commented on build.gradle:
```diff
@@ -12,6 +12,12 @@ repositories {
     testImplementation platform('org.junit:junit-bom:5.9.1')
     testImplementation('org.junit.jupiter:junit-jupiter')
+
+    // Assertj
+    testImplementation platform('org.junit:junit-bom:5.9.1')
+    testImplementation platform('org.assertj:assertj-bom:3.25.1')
+    testImplementation('org.junit.jupiter:junit-jupiter')
+    testImplementation('org.assertj:assertj-core')
```
요구사항 👍

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
-                validateOverflow(a, b, operator);
-                validateUnderflow(a, b, operator);
+    private List<Integer> convertInputIntoList(String input, String separator) {
+        List<Integer> numbers = new ArrayList<>();
+
+        // 커스텀 구분자일 경우, 커스텀 구분자 선언 부분을 자른다
+        if (isCustomSeparator(separator)) {
+            input = input.substring(CUSTOM_SEPARATOR_STARTER.length() + CUSTOM_SEPARATOR_TERMINATOR.length() + separator.length());
+        }
+
+        StringTokenizer tokenizer = new StringTokenizer(input, separator);
+        while (tokenizer.hasMoreTokens()) {
+            String token = tokenizer.nextToken();
+
+            if (!isDigit(token)) {
+                throw new RuntimeException("숫자 이외의 값은 전달할 수 없습니다.");
             }
-            case DIVIDE -> validateDivide(b);
+            numbers.add(Integer.parseInt(token));
         }
+
+        return numbers;
     }
```
### 불변성(Immutable) 유지의 중요성

현재 메서드는 `List<Integer> 객체의 참조`를 반환하고 있습니다. 이렇게 반환된 리스트가 다른 곳에서 수정될 가능성이 있다면, 예기치 않은 부작용이 발생할 수 있어요. 예를 들어, 반환된 리스트가 외부에서 변경된다면 원래 기대했던 결과와 달라질 수 있겠죠?

이를 방지하려면 불변 리스트를 반환하거나, 복사본을 만들어 반환하는 방법을 고려할 수 있을 것 같습니다!

`Collections.unmodifiableList(numbers);` 또는, `List.copyOf(numbers);` 등 이렇게 하면 외부에서 리스트(객체)를 수정하는 것을 방지할 수 있어요.

### 매개변수 수정의 문제점

현재 매개변수로 받은 input 문자열을 직접 변경하고 있습니다. 이는 입력값이 메서드 내부에서 변경되면서 코드의 가독성과 유지보수성을 떨어뜨리는 원인이 될 수 있을 것 같아요.

왜 문제가 될까요?

우선 매개변수의 상태가 변하면서 코드 이해가 어려워집니다. 

처음에는 **"전체 입력값"**을 의미했지만, 중간에 **"커스텀 구분자를 제거한 값(식에 해당하는 문자열)"**으로 바뀌어요. 이렇게 상태가 변하면 유지보수할 때 혼란을 줄 수 있어요. (또한 커스텀 구분자가 있냐 없냐에 따라 이 상태의 가짓수는 더욱 많아져요)

### 단일 책임 원칙(SRP) 위반 가능성

하나의 메서드가 `커스텀 구분자를 처리하는 역할`과 `문자열을 정수 리스트로 변환하는 역할`을 동시에 수행하고 있어요. 이런 경우에는 두 가지 역할을 분리하면 코드가 더 명확해질 것 같아요. 커스텀 구분자를 처리하는 메서드를 만들고 그 반환 값을 해당 메서드에 넘겨 식에 해당하는 문자열을 정수형 리스트에 담아 반환하는 역할을 수행하도록 하면 더 좋을 것 같습니다.

kjoon418 replied:
> 불변성(Immutable) 유지의 중요성

상수를 불변 컬렉션으로 선언하는 정도만 하고 있었는데, 메서드의 반환값 역시 불변인 것이 좋겠군요!
불변 컬렉션을 반환한다는 목적을 명확하게 하기 위해 `Collections.unmodifiableList(numbers)` 메서드를 사용하도록 하겠습니다!

> 매개변수 수정의 문제점

커스텀 구분자를 사용하는 `input`은 `//{구분자}\n{숫자}` 형식을 지니고, 그렇지 않은 input은 `{숫자}` 형식을 지니므로 둘을 통일시키려 했습니다.
그러나 그 과정에서 매개변수를 수정하게 되면서 코드의 가독성과 유지보수성을 떨어뜨리는군요..!
로직을 수정하고, 변수의 이름을 현재 상태에 걸맞게 직관적으로 지어주도록 하겠습니다!

> 단일 책임 원칙(SRP) 위반 가능성

메서드의 구조가 복잡하다 느껴졌었는데, 단일 책임 원칙을 준수하지 않았기 때문이군요.
말씀해주신 것 처럼, `input에서 커스텀 구분자 선언부를 제거하는 메서드`와 `문자열을 리스트로 변환하는 메서드`를 각각 만들도록 하겠습니다!

피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+        return number > 0;
+    }
+
+    private static boolean isCustomSeparator(String separator) {
+        return !BASIC_SEPARATORS.contains(separator);
+    }
```
기본 구분자가 아니라는 것만으로 커스텀 구분자라고 할 수 있을까요?

예를 들어, "3"을 전달하면 "3"은 커스텀 구분자가 될 수 있나요? 그렇다면 식에서 333은 어떻게 해석할 수 있을까요?

> 다른 곳에서 숫자는 커스텀 구분자로 사용할 수 없게 막을 수도 있겠지만, 해당 메서드를 사용하려는 개발자에게는 명확하지 않은 메서드로 인식될 수 있을 것 같아요.

kjoon418 replied:
커스텀 구분자인지 확인하는 역할을 수행하게 하려면 추가적인 검증 로직이 필요하겠군요!
(혹은, 조건문을 유지하고 싶다면 메서드 이름을 `isNotBasicSeparator`로 수정해야 하고요)

해당 메서드를 구상할 당시에는 숫자 역시 커스텀 구분자로 사용할 수 있게 하려 했는데, 말씀해주신 예시 상황과 같은 문제가 발생할 수 있겠군요.
다양한 예외 상황에 대처할 수 있도록 로직을 수정하도록 하겠습니다.

피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
-                if ((long) a - (long) b > Integer.MAX_VALUE) {
-                    throw new IllegalArgumentException(String.format("(%d) - (%d)는 오버플로우가 발생합니다.", a, b));
-                }
-            }
-            case MULTIPLY -> {
-                if ((long) a * (long) b > Integer.MAX_VALUE) {
-                    throw new IllegalArgumentException(String.format("(%d) * (%d)는 오버플로우가 발생합니다.", a, b));
-                }
+        }
+
+        throw new RuntimeException("적절한 구분자를 찾지 못했습니다.");
+    }
```
메서드의 역할이 뚜렷하며, 기본 구분자를 찾는 기능에 집중하고 있다고 느껴집니다. 👍 

찾지 못하는 경우 예외를 바로 던지는 것도 괜찮을 수 있지만, 해당 메서드를 사용하는 클라이언트(자신, 팀원 등)가 더 유연하게 사용할 수 있도록 Optional을 사용하는 것도 괜찮을 것 같아요.

kjoon418 replied:
말씀해주신 것 처럼, 현재 메서드는 기본 구분자를 찾지 못할 시 항상 예외를 던지므로 클라이언트 단에서 유연하게 사용할 수 없을 것 같습니다.
당장의 로직에서는 바로 예외를 던지는 것이 간결해 보여서 위와 같이 작성했지만, 앞으로는 메서드의 유연성과 확장성까지 고려하도록 하겠습니다.

피드백 감사합니다!

YehyeokBang commented on src/test/java/CalculatorTest.java:
```diff
@@ -397,7 +396,187 @@ void negative_negative() {
+                    .isInstanceOf(RuntimeException.class);
+        }
+
+        @Test
+        @DisplayName("커스텀 구분자 선언부를 작성하지 않을시 예외가 발생한다")
+        void notDeclareCustomSeparator() {
+            // given
+            String input = "1$2$3$4$5$6$7$8$9$10";
+
+            // then
+            assertThatThrownBy(() -> calculator.addAll(input))
+                    .isInstanceOf(RuntimeException.class);
+        }
```
커스텀 구분자를 선언하지 않아서 예외가 발생했다기 보다는 알 수 없는(선언되지 않은) 커스텀 구분자를 사용해서 발생한 문제에 가깝지 않을까요? 어떻게 생각하시나요?

kjoon418 replied:
맞습니다. 기본 구분자만을 사용할 때 역시 커스텀 구분자 선언부를 작성하지 않는데, 이 경우에는 문제가 발생하지 않습니다.
그러므로 문제의 근본적인 원인은 '선언되지 않은 커스텀 구분자 사용'이므로, 이에 맞게 `@DisplayName`을 수정하도록 하겠습니다.

피드백 감사합니다!

YehyeokBang commented on src/test/java/CalculatorTest.java:
```diff
@@ -397,7 +396,187 @@ void negative_negative() {
+                    .isInstanceOf(RuntimeException.class);
+        }
+
+        @Test
+        @DisplayName("실수 전달시 예외가 발생한다")
+        void floatException() {
+            // given
+            String input = "1,2,3.3,4,5,6,7,8,9,10";
+
+            // then
+            assertThatThrownBy(() -> calculator.addAll(input))
+                    .isInstanceOf(RuntimeException.class);
+        }
+    }
```
`.`을 커스텀 구분자로 사용하면 어떻게 되나요?

kjoon418 replied:
![image](https://github.com/user-attachments/assets/3aa70352-dab9-49e1-887c-250759779db5)
기능 요구사항의 예시를 보고, 정수형 값에 대한 계산기를 만들라는 것으로 이해했습니다.
그러므로 해당 메서드는 실수 전달을 허용하지 않으니, `.`은 커스텀 구분자로 사용할 때만 허용되도록 설계했습니다.

하지만 다시 확인해 보니 정수형 값만 받으라는 요구사항이 명시되어 있지 않네요.
우선은 정수형 값만 받도록 설계하되, 요구사항 변경시 코드를 조금만 수정해도 실수형 값을 받을 수 있도록 구현하겠습니다!

또한 [정규 표현식 문법](https://gbminnote.com/entry/%EC%A0%95%EA%B7%9C%ED%91%9C%ED%98%84%EC%8B%9DRegular-Expression-%EC%A0%95%EC%9D%98%EC%99%80-%EA%B0%9C%EB%85%90#:~:text=%EC%A0%95%EA%B7%9C%20%ED%91%9C%ED%98%84%EC%8B%9D%20%28Regular%20Expression%29%EC%9D%80%20%EB%AC%B8%EC%9E%90%EC%97%B4%EC%9D%98%20%ED%8C%A8%ED%84%B4%EC%9D%84%20%EC%B0%BE%EC%95%84%EB%82%B4%EA%B3%A0%2C%20%EA%B7%B8,%EA%B0%80%EC%A7%80%EB%8A%94%20%ED%8C%A8%ED%84%B4%20%EB%93%B1%EC%9D%84%20%EC%B0%BE%EC%95%84%EB%82%B4%EB%8A%94%20%EB%8D%B0%20%ED%99%9C%EC%9A%A9%EB%90%A0%20%EC%88%98%20%EC%9E%88%EB%8B%A4.)을 찾아보니, `.`은 줄 바꿈 문자를 제외한 모든 문자와 매칭되는 메타문자로 사용되네요.
`.` 뿐 아니라 `^`, `$` 역시 정규 표현식에서 메타문자로 사용되는데, 이와 같은 메타문자는 커스텀 구분자로 사용할 수 없도록 제약해야 할지, 메타문자도 커스텀 구분자로 사용할 수 있도록 로직을 작성(메타문자 앞에 `\`를 추가)해야 할지 고민이 됩니다.
우선은 커스텀 구분자로 메타문자를 사용하려 할 경우 앞에 `\`를 추가하도록 코드를 수정해보겠습니다!

YehyeokBang replied:
스스로 판단하고 이유를 가진채로 새로운 문제를 해결하려는 모습이 좋은 것 같습니다. 메타 문자 외에도 수학적 기호에 대해서도 고민해보면 좋을 것 같아요! 

> 예를 들어 커스텀 구분자를 `-`로 사용하고 `"20-2-2" 문자열을 사용했을 때, 누군가는 (실제 빼기 연산으로) 16을 결과로 기대할 수 있지만, 실제로는 모두 더하기 연산으로 24가 반환되어야 하기 때문이에요. 

콘솔 또는 UI로 프로그램을 사용할 때 사용자가 경험할 순간까지 생각하면 더 견고한 프로그램을 만들 수 있을 것 같아요! (물론 항상 완벽한 프로그램을 만들 수는 없지만요..😢)

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+        return isBiggerThanZero(endIndex - CUSTOM_SEPARATOR_STARTER.length());
+    }
+
+    private static boolean isDigit(String string) {
+        for (char word : string.toCharArray()) {
+            if (!Character.isDigit(word)) {
+                return false;
             }
         }
+
+        return true;
     }
```
전달하는 문자열이 빈 문자열이라면 어떻게 될까요?

kjoon418 replied:
문자열이 비어 있을 경우를 고려하지 못했군요.
해당 메서드가 동작할 때 어떤 예외 케이스들이 있을지 항상 생각하도록 하겠습니다.
피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+        // 시작점과 끝점이 존재하는지 확인한다
+        int startIndex = string.indexOf(CUSTOM_SEPARATOR_STARTER);
+        int endIndex = string.indexOf(CUSTOM_SEPARATOR_TERMINATOR);
+        if (isNegative(startIndex) || isNegative(endIndex)) {
+            return false;
+        }
```
조건을 이해하기 수월했습니다. 👍

kjoon418 replied:
메서드 이름을 통해 조건을 직관적으로 이해할 수 있어, 코드의 가독성이 향상된게 느껴졌습니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+            return false;
+        }
+
+        // 시작점과 끝점 형식이 일치하는지 확인한다
+        String actualStarter = string.substring(startIndex, CUSTOM_SEPARATOR_STARTER.length());
+        String actualTerminator = string.substring(endIndex, endIndex + CUSTOM_SEPARATOR_TERMINATOR.length());
+        if (!actualStarter.equals(CUSTOM_SEPARATOR_STARTER) || !actualTerminator.equals(CUSTOM_SEPARATOR_TERMINATOR)) {
+            return false;
+        }
+
+        // 커스텀 구분자의 길이가 1을 넘을 경우에만 올바르다 판단한다
+        return isBiggerThanZero(endIndex - CUSTOM_SEPARATOR_STARTER.length());
```
주석 없이도 이해하기 쉬운 코드로 만들어보는 것은 어떨까요? `올바르게 (형식에 맞게) 커스텀 구분자를 선언했는지?` 라는 내용으로 메서드를 분리하는 것도 좋은 방법이 될 것 같아요.

kjoon418 replied:
```java
        String actualStarter = string.substring(startIndex, CUSTOM_SEPARATOR_STARTER.length());
        String actualTerminator = string.substring(endIndex, endIndex + CUSTOM_SEPARATOR_TERMINATOR.length());
        if (!actualStarter.equals(CUSTOM_SEPARATOR_STARTER) || !actualTerminator.equals(CUSTOM_SEPARATOR_TERMINATOR)) {
            return false;
        }
```

말씀해주신대로 코드를 수정하려 하니, 위 조건문이 불필요한 것을 발견했습니다.
불필요한 코드를 제거하고 주석 없이 이해할 수 있도록 개선하도록 하겠습니다.
피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+        return isBiggerThanZero(endIndex - CUSTOM_SEPARATOR_STARTER.length());
+    }
+
+    private static boolean isDigit(String string) {
+        for (char word : string.toCharArray()) {
+            if (!Character.isDigit(word)) {
+                return false;
             }
         }
+
+        return true;
     }
```
Character의 정적 메서드인 isDigit()과 이름이 같습니다. 이 경우에 어떤 문제가 발생할 수 있을까요?

kjoon418 replied:
파라미터로 `String`을 받는 것으로 문자열에 대한 `isDigit()` 메서드라 선언해, `Character.isDigit()`과 구분하고자 했습니다.
하지만 이 경우 `Character`를 `static import` 하게 될 경우 메서드 이름이 겹치게 되어 어떤 메서드를 사용하는지 구분하기 힘들 수 있겠다는 생각이 드네요.
또한 `Calculator` 클래스 안에 `isDigit()`이 있는 것을 모르는 개발자가 있을 경우, 해당 메서드를 사용하는 코드를 해석하기 어려울 것 같습니다.

현재 위의 메서드는 "123"과 같이 정수 숫자로만 구성된 문자열에 한해서만 true를 반환하고 있으니, 메서드의 이름을 `isIntegerString()`이라고 수정하겠습니다!

YehyeokBang replied:
발생 가능한 문제까지 잘 파악하신 것 같습니다.👍

참고로 `isNumeric()`도 괜찮은 선택이 될 수 있을 것 같아요. `isIntegerString()`과 비교해보는 것도 좋을 것 같네요.

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+     * @throws RuntimeException 구분자가 부적절함<br>숫자 이외의 값이 전달됨<br>음수가 전달됨
+     */
+    public int addAll(String input) {
+        String separator = getSeparatorFromInput(input);
+        List<Integer> numbers = convertInputIntoList(input, separator);
 
-        return a * b;
+        int sum = 0;
+        for (Integer number : numbers) {
+            sum += number;
+        }
+
+        return sum;
     }
```
한 번의 계산에는 하나의 구분자만 사용할 수 있는 것 같은데 이해한게 맞을까요?

![image](https://github.com/user-attachments/assets/e0052484-7378-47d3-8e75-664ab4a87f4c)

저는 3단계 힌트를 보고나서 기본 구분자와 (선언한 경우 커스텀 구분자까지) 모두 하나의 계산 안에 사용할 수 있도록 의도된 것 같아서 여쭤봅니다! 

> 제가 생각한 예시
> //@\n1:2@3,4 가능

kjoon418 replied:
이해하신 대로 한 번의 계산에는 하나의 구분자만 사용하도록 설계했습니다.
말씀하신 것 처럼 기본 구분자와 (선언한) 커스텀 구분자까지 한 계산 안에 사용할 수 있도록 로직을 수정하겠습니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+
+        abstract int operate(int leftOperand, int rightOperand);
+
+        private static void validateOutOfRangeIfAdd(int a, int b) {
+            if (isWithoutIntRange((long) a + (long) b)) {
+                throw new IllegalArgumentException(String.format("(%d)와(과) (%d) 연산시 오버플로우 혹은 언더플로우가 발생합니다.", a, b));
             }
         }
```
매번 long 타입으로 변환하지 않고 오버플로우(또는 언더플로우)를 감지할 수 있는 방법은 없을까요?

kjoon418 replied:
음... 좋은 방법이 떠오르지 않네요. 몇 가지 방법은 생각이 났으나, 그리 좋은 방법은 아닌 것 같습니다.
1. 피연산자와 연산 결과값의 부호가 다른지를 검사한다.
   - 양수와 양수를 더했는데 결과가 음수거나, 그 반대의 경우(`음수 + 음수 = 양수`) 오버플로우 혹은 언더플로우가 발생했다고 추측할 수 있습니다.
   - 하지만 두 피연산자가 모두 양수거나 모두 음수일 경우에만 적용해야 하니 조건이 복잡해질 수 있고, 그로 인해 코드를 읽는 입장에서 가독성이 떨어질 것 같습니다.
     (물론 덧셈에서 두 피연산자의 부호가 다를 경우엔 오버플로우/언더플로우가 발생할 일이 없긴 합니다.)

어떻게 개선할 수 있는지 조언해 주신다면 정말 감사하겠습니다!

YehyeokBang replied:
```java
if ((b > 0 && a > Integer.MAX_VALUE - b) || (b < 0 && a < Integer.MIN_VALUE - b)) {
        // 예외 던지기
}
```

- 덧셈 연산을 수행하기 전에 미리 최대/최소값과 비교하여 감지하는 방법도 있을 것 같아요!

kjoon418 replied:
이런 방법도 있었군요!
해당 메서드 내부 로직을 이해하긴 조금 더 어려울 순 있어도, 해당 메서드를 사용할 클라이언트 측에선 단순히 `int` 값 2개를 전달해주면 되니 편리할 것 같습니다.
또한, 유틸 메서드는 이름을 잘 지어주는 것으로 의미를 표현할 수 있으니 내부 로직이 다소 어려워도 큰 문제가 되진 않겠네요.
조언 감사합니다!!

YehyeokBang commented on src/test/java/CalculatorTest.java:
```diff
@@ -397,7 +396,187 @@ void negative_negative() {
+        return sum;
+    }
+
+    enum Separator {
+        BASIC {
+            @Override
+            String convertListIntoString(List<Integer> numbers, String separator) {
+                StringBuilder builder = new StringBuilder();
+
+                for (Integer number : numbers) {
+                    builder.append(number)
+                            .append(separator);
+                }
+
+                return builder.delete(builder.length() - separator.length(), builder.length())
+                        .toString();
+            }
+        },
+        CUSTOM {
+            @Override
+            String convertListIntoString(List<Integer> numbers, String separator) {
+                StringBuilder builder = new StringBuilder()
+                        .append(CUSTOM_SEPARATOR_STARTER)
+                        .append(separator)
+                        .append(CUSTOM_SEPARATOR_TERMINATOR);
+
+                for (Integer number : numbers) {
+                    builder.append(number)
+                            .append(separator);
+                }
+
+                return builder.delete(builder.length() - separator.length(), builder.length())
+                        .toString();
+            }
+        };
+
+        abstract String convertListIntoString(List<Integer> numbers, String separator);
     }
```
무슨 역할을 수행하는지 쉽게 파악하기 어려운 것 같습니다. 해당 코드의 의도를 설명해주실 수 있나요?

kjoon418 replied:
해당 enum은 테스트의 중복 코드를 제거하기 위해, `List를 addAll()의 형식에 맞는 문자열로 변환`하는 역할을 합니다.
그러나 `Separator`, `BASIC`, `CUSTOM`, `convertListIntoString` 모두 해당 역할에 어울리지 않는 이름을 지니고 있어 다른 개발자 입장에서 해당 코드가 어떤 역할을 수행하는지 파악이 어려운 것 같습니다.
역할에 맞는 이름을 지니도록 수정하도록 하겠습니다.

피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,205 @@
+        return !BASIC_SEPARATORS.contains(separator);
+    }
+
+    enum Operator {
```
enum 활용 👍

kjoon418 replied:
리뷰어님의 피드백 덕분에 `enum` 활용법을 공부할 수 있었습니다.
감사합니다!

YehyeokBang reviewed (changes requested):
3, 4단계 미션 고생하셨습니다! 👍 

저번 미션보다 메서드 단위로 읽기 쉬워진 것 같습니다. 추가로 드리고 싶은 말씀은 리뷰로 작성해 두었으니 편하게 답변해 주세요. 😄 

> 피드백해 주신 대로 조건문을 메서드로 추출하여 의미를 부여했습니다.
가독성은 좋아진 대신 클래스에 메서드가 너무 많아졌는데, 이 부분은 크게 문제되지 않나요?

- 어떤 형태의 문제를 의식하고 말씀하신 것인지 더 명확하게 추가 설명을 부탁드려도 될까요?
  - 단순히 하나의 클래스가 길어지는 것, 매번 메서드를 확인할 때 아래로 내려가서 확인해야 하는 것 등
  
> 동작을 설명하기 위해 Enum의 추상 메서드 이름을 operate, convertListIntoString으로 지었습니다.
함수형 인터페이스 관례에 따라 apply로 짓는 것이 더 나을까요?

- 저는 지금도 괜찮을 것 같아요.

> 단계의 요구사항을 수행하는 메서드를 만드는데, 메서드 이름만으로 동작을 직관적으로 설명하기 어려워 주석을 통해 문서화했습니다.
좋은 문서화가 아닌 것 같은데, 어떤 방법으로 개선할 수 있을까요?

- 하나의 메서드가 가지는 책임이 여러 개인 경우가 많아서 그것들을 메서드명으로 표현하기 어려웠던 것이 아닐까? 라는 하나의 의견을 남기겠습니다..! 정말 하나의 일만을 수행하고 있는지 생각해보시면 더 나은 메서드명으로 작성할 수 있을 것 같아요. 😄 

> 문자열을 나눌 때 기존에 자주 사용하던 StringTokenizer를 사용했습니다.

- 미션을 수행할 때는 요구사항을 준수하는 것을 우선순위로 둔다면 더 좋을 것 같습니다!
  - 미션 이후에 함께 미션을 수행한 사람들과 그런 내용으로 이야기를 해본다면 좋을 것 같네요.
- 제가 미션을 출제한 것이 아니라 내부 의도까지는 알기 어려운 것 같습니다.

> 커스텀 구분자의 시작점(//)과 끝점(\n)을 상수로 표현했습니다.
그러나 해당 상수가 Calculator와 CalculatorTest 모두에 필요한 상황인데 어떻게 처리하는 것이 바람직할까요?

- 항상 정답은 없는 것 같습니다. 그러나 더욱 성장하고 싶은 마음에 정답(어떻게 보면 좋은 코드)에 가까운 코드를 찾으려는 마음은 누구나 생기는 것 같아요. 
- `나열해두고 무엇이 더 나을까요?` 보다는 `현재 상황에서는 이 방법의 장점이 크게 느껴지고, 다른 방법의 단점을 충분히 상쇄할 수 있는 것 같습니다.`와 같은 방향으로 질문을 남겨주시면 더 좋은 리뷰를 할 수 있을 것 같습니다!
- 저는 테스트를 위한 접근 지정자 마지노선은 protected까지로 생각하고 있습니다. 가독성을 위해 유틸 클래스나 Enum을 사용하는 것도 괜찮을 것 같네요.

> 깔끔하게 변경 사항 확인하기

![image](https://github.com/user-attachments/assets/94bdae3f-4270-4811-82c0-ffe64fbb46da)

- Review Changes를 눌러서 설정을 변경하면 보다 깔끔하게 확인할 수 있습니다!
- 코드 리뷰에 더욱 집중할 수 있도록 이러한 질문은 구글링을 통해 먼저 찾아주시고 해결하기 어려운 경우에 남겨주시면 감사하겠습니다. 서로의 시간은 소중하고 더 나은 리뷰를 작성해드리고 싶은 마음에 드리는 말씀이니 양해를 부탁드립니다..! 😄 


---

### 추가 내용

<img width="753" alt="image" src="https://github.com/user-attachments/assets/dd1654e9-144c-4360-879b-74e833dee5e3" />

프로그래밍 요구 사항인 만큼 둘의 차이를 간단하게 생각해보면 좋을 것 같습니다!

kjoon418 committed:
refactor: 피드백 반영

범용적으로 사용할 유틸 메서드들을 별도의 클래스에 배치함.
여러 클래스에서 같이 사용할 상수값을 별도의 클래스에 배치함.
요구사항을 준수하도록 동작을 수정함.
피드백대로 코드를 개선함.

kjoon418 committed:
fix: 누락된 클래스들을 추가함

kjoon418 committed:
fix: addAll 메서드 문서화 내용 수정

kjoon418 committed:
fix: 정규 표현식 관련으로 문제를 일으키던 코드 수정

kjoon418 committed:
refactor: 내부 클래스로 구현된 기능을 메서드로 옮김

kjoon418 commented:
피드백해 주신대로 코드를 수정하고 `push`했습니다!
또한, 코멘트에 대한 답변과 추가 질문을 게시했습니다.

## 답변

> 피드백해 주신 대로 조건문을 메서드로 추출하여 의미를 부여했습니다.
가독성은 좋아진 대신 클래스에 메서드가 너무 많아졌는데, 이 부분은 크게 문제되지 않나요?
>> 어떤 형태의 문제를 의식하고 말씀하신 것인지 더 명확하게 추가 설명을 부탁드려도 될까요?
단순히 하나의 클래스가 길어지는 것, 매번 메서드를 확인할 때 아래로 내려가서 확인해야 하는 것 등

- 포괄적으로 질문해 죄송합니다.
  말씀해주신 것 처럼, 기존에 비해 한 클래스의 길이가 과하게 길어진다는 느낌을 받았습니다.
- 이로 인해 다른 개발자가 해당 클래스의 구조를 이해하기 어려워질까 우려가 되었으며, 저 또한 유지보수 과정에서 원하는 메서드가 어디 있는지 찾기 힘들었습니다.
  (IntelliJ의 도움으로 `Ctrl + 좌클릭`을 통해 해당 메서드로 곧바로 찾아갈 순 있지만, `A 메서드를 수정해야 할 것 같은데 어디있지?`와 같은 상황에는 시간이 많이 소요되었습니다.)
- 우선은 여러 클래스에서 사용해도 무방한 유틸 메서드는 별도의 유틸 클래스에 배치했습니다.
  - 다만, 추후에 여러 클래스에서 한 유틸 클래스에 접근하게 되어, 해당 유틸 클래스를 수정할 시 영향을 미치는 범위가 매우 커질 수 있는 것이 우려스럽습니다.

> 문자열을 나눌 때 기존에 자주 사용하던 StringTokenizer를 사용했습니다.
>> 미션을 수행할 때는 요구사항을 준수하는 것을 우선순위로 둔다면 더 좋을 것 같습니다!
미션 이후에 함께 미션을 수행한 사람들과 그런 내용으로 이야기를 해본다면 좋을 것 같네요.
제가 미션을 출제한 것이 아니라 내부 의도까지는 알기 어려운 것 같습니다.

- 해당 페이지가 요구사항이 아닌 단순 힌트라고 판단해, 저에게 더 익숙한 방식을 사용했습니다.
  요구사항을 준수하지 않아 죄송합니다!
- `String.split()`을 사용하도록 코드를 수정했습니다.

> 커스텀 구분자의 시작점(//)과 끝점(\n)을 상수로 표현했습니다.
그러나 해당 상수가 Calculator와 CalculatorTest 모두에 필요한 상황인데 어떻게 처리하는 것이 바람직할까요?
>> 나열해두고 무엇이 더 나을까요? 보다는 현재 상황에서는 이 방법의 장점이 크게 느껴지고, 다른 방법의 단점을 충분히 상쇄할 수 있는 것 같습니다.와 같은 방향으로 질문을 남겨주시면 더 좋은 리뷰를 할 수 있을 것 같습니다!
저는 테스트를 위한 접근 지정자 마지노선은 protected까지로 생각하고 있습니다. 가독성을 위해 유틸 클래스나 Enum을 사용하는 것도 괜찮을 것 같네요.

- 말씀해주신 내용 명심하여, 이후로는 더 좋은 방식으로 질문할 수 있도록 노력하겠습니다.
  잘못된 방법으로 질문 드려 죄송합니다!
- 우선은 별도의 유틸 클래스를 만들어 해결했습니다.
  Enum도 고려하였으나, 상수의 자료형이 `Set`, `String`으로 다양했기에 클래스로 만드는 것이 코드가 간결해질 것이라 판단했습니다.

> Review Changes를 눌러서 설정을 변경하면 보다 깔끔하게 확인할 수 있습니다!
코드 리뷰에 더욱 집중할 수 있도록 이러한 질문은 구글링을 통해 먼저 찾아주시고 해결하기 어려운 경우에 남겨주시면 감사하겠습니다. 서로의 시간은 소중하고 더 나은 리뷰를 작성해드리고 싶은 마음에 드리는 말씀이니 양해를 부탁드립니다..! 😄

- PR 단위를 잘 나누어서 해결해야 하는 문제인지 알았는데, 깃허브 기능으로 해결하는 것이었군요...!
  부적절한 질문 드려 죄송합니다. 이후로는 꼭 구글링을 통해 해결할 수 있는 문제인지 확인한 후 질문 드리도록 하겠습니다!

> 프로그래밍 요구 사항(JUnit5 -> AssertJ)인 만큼 둘의 차이를 간단하게 생각해보면 좋을 것 같습니다!

- 기존에 AssertJ를 사용해와서, 역으로 JUnit5를 사용해 테스트 코드를 작성할 때 불편함을 느꼈던 것 같습니다.
- AssertJ의 경우, 코드를 작성할 때 의식의 흐름에 맞게 `실제 값(A)이 -> 예상한 값(B)과 일치하는가?`를 메서드 체이닝을 사용해 보기 좋게 작성할 수 있던 반면, JUnit5의 경우는 `A, B가 일치하는가?`처럼 작성하게 되는 것 같아서 직관성이 떨어진다고 느꼈습니다.
- 그와 별개로, 여러 케이스에 대한 테스트를 작성하니 클래스를 수정했을 때 실수를 빠르게 파악하고 수정할 수 있어 굉장한 편리함을 느낄 수 있었습니다.

## 추가 질문
### 1. 문서화
```java
    /**
     * 구분자를 가진 문자열을 받아 모든 숫자의 합을 반환한다. <br>
     * 쉼표 또는 콜론을 기본 구분자로 사용한다. <br>
     * 커스텀 구분자를 사용하고 싶을 경우, 문자열 앞부분에 "//${커스텀구분자}\n"을 추가한다. <br>
     * ex_ "//;\n1;2;3" <br>
     *
     * @param input 구분자로 구분된 자연수 양수를 지닌 문자열
     * @return 모든 숫자의 합
     * @throws RuntimeException 구분자가 부적절함<br>숫자 이외의 값이 전달됨<br>음수가 전달됨
     */
    public int addAll(String input) {
            // ...
    }
```
- 3단계 요구사항을 수행하는 메서드를 구현하는데, 메서드의 이름을 잘 짓는 것 만으로는 클라이언트 측에서 해당 메서드를 이해하기 힘들 것이라 생각했습니다.
- 그래서 위와 같이 주석을 사용해 문서화 했는데, 문서화 방식에 문제는 없을까요?
  (컨벤션에 맞지 않거나, 놓친 설명이 있거나 등등)

### 2. 정규 표현식 처리
- 문제의 요구사항에 정확히 명시되어 있지 않지만, `.`과 `$`같이 정규 표현식에서 메타문자로 사용하는 값 역시 커스텀 구분자로 사용할 수 있어야 한다고 판단했습니다.
- 현재는 커스텀 구분자로 메타문자를 사용할 경우, 해당 메타문자 앞에 이스케이프를 붙여 일반 문자처럼 사용되도록 했습니다.
- 수정한 로직 역시 과제의 요구사항에 부적합한지 궁금합니다.
- 정규 표현식과 관련해 [해당 자료](https://gbminnote.com/entry/%EC%A0%95%EA%B7%9C%ED%91%9C%ED%98%84%EC%8B%9DRegular-Expression-%EC%A0%95%EC%9D%98%EC%99%80-%EA%B0%9C%EB%85%90#:~:text=%EC%A0%95%EA%B7%9C%20%ED%91%9C%ED%98%84%EC%8B%9D%20%28Regular%20Expression%29%EC%9D%80%20%EB%AC%B8%EC%9E%90%EC%97%B4%EC%9D%98%20%ED%8C%A8%ED%84%B4%EC%9D%84%20%EC%B0%BE%EC%95%84%EB%82%B4%EA%B3%A0%2C%20%EA%B7%B8,%EA%B0%80%EC%A7%80%EB%8A%94%20%ED%8C%A8%ED%84%B4%20%EB%93%B1%EC%9D%84%20%EC%B0%BE%EC%95%84%EB%82%B4%EB%8A%94%20%EB%8D%B0%20%ED%99%9C%EC%9A%A9%EB%90%A0%20%EC%88%98%20%EC%9E%88%EB%8B%A4.)를 참고했습니다.

### 3. 로깅
- 정규 표현식 관련 로직을 구현하는데 많은 시행착오를 겪어, 이를 빠르게 파악하고자 로그를 추가했습니다.
- 평소에는 스프링 프로젝트에서 Slf4J 라이브러리의 로거를 사용했는데, 해당 과제에서는 Java가 기본적으로 제공해주는 로거를 사용하고자 `java.util.logging.Logger`를 사용했습니다.
- 제가 작성한 로그 코드에 개선점이 있을지 궁금합니다.(불필요하게 사용하고 있거나, 알아보기 어렵거나 등등)

kjoon418 committed:
chore: 테스트 케이스 추가

YehyeokBang commented:
### 참고

리뷰 반영 및 답변 이후에 리뷰를 다시 요청하고 싶을 때 PR 오른쪽 상단에 리뷰어 목록에서 지정하여 다시 요청을 보낼 수 있습니다! (누르면 주황색 불로 바뀝니다.) @kjoon418 

<img width="943" alt="image" src="https://github.com/user-attachments/assets/3833b3c5-1caa-4a91-98b0-bb31e0e4fcd6" />

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,233 @@
+            numbers = convertStringNumbersIntoList(input, BASIC_REGEX);
+        }
+
+        int sum = 0;
+        for (Integer number : numbers) {
+            sum += number;
+        }
```
### 생각해보면 좋을 부분

`sum += number;` 구문에서 `number`는 `Integer` 타입이고 `sum`은 `int` 타입입니다. 이때 어떻게 연산이 될까요? 반복문을 돌면서 매번 `언박싱`이 발생한다면, 불필요한 연산이 추가되어 성능 저하가 발생할 수 있는 코드일까요?

kjoon418 replied:
```suggestion
        Integer sum = 0;
        for (Integer number : numbers) {
            sum += number;
        }
```
컬렉션에 담긴 primitive type은 자동으로 Wrapper 클래스로 변환되므로, int 타입인 sum에 더해질때마다 언박싱이 발생해 성능 저하가 일어나겠군요...!
위와 같이 수정하는 것이 바람직해 보이네요.
피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,233 @@
+    private static String createRegexByCustomSeparator(String customSeparator) {
+        StringBuilder regexBuilder = new StringBuilder();
+
+        for (String basicSeparator : BASIC_SEPARATORS) {
+            regexBuilder.append(basicSeparator)
+                    .append("|");
+        }
```
다른 개발자가 봤을 때 `"|"`를 이해하기 어려울 수 있을 것 같아요. 상수로 처리하는 방법에 대해서는 어떻게 생각하시나요?

kjoon418 replied:
이 부분 역시 상수를 사용할 수 있었군요!
"또는"이라는 의미를 잘 나타낼 수 있는 상수명을 고민해 보겠습니다.
피드백 감사합니다!

YehyeokBang commented on src/main/java/Calculator.java:
```diff
@@ -1,86 +1,233 @@
 
-    public int add(int a, int b) {
-        validateOperation(a, b, Operator.ADD);
+    private static final Set<String> BASIC_SEPARATORS = CalculatorConstant.BASIC_SEPARATORS;
+    private static final String CUSTOM_SEPARATOR_STARTER = CalculatorConstant.CUSTOM_SEPARATOR_STARTER;
+    private static final String CUSTOM_SEPARATOR_TERMINATOR = CalculatorConstant.CUSTOM_SEPARATOR_TERMINATOR;
+    private static final Set<Character> REGEX_META_CHARACTERS = CalculatorConstant.REGEX_META_CHARACTERS;
```
상수를 그대로 사용하지 않고 한번 더 선언한 이유가 있나요?

kjoon418 replied:
기존 코드를 수정하는 과정에서 위와 같이 작성한 것이었는데, `CalculatorConstant`를 `static import`해서 바로 사용하는 것이 더 바람직하겠군요.
실수를 바로잡아 주셔서 정말 감사합니다!

YehyeokBang reviewed (approved):
> 조건문을 메서드로 추출하여 가독성을 좋아졌지만, 한 클래스의 길이가 과하게 길어짐. 스스로도 코드를 찾을 때 오랜 시간이 걸렸다. 

단순히 메서드가 많아지는 현상만을 보고서 이것이 문제라고 인식하기 부족하다고 생각합니다. 해결해야할 문제가 실제로도 복잡하다면 메서드는 더 많아질 수 있을 것 같아요.

- 우선 클래스의 책임이 무겁지 않은가를 먼저 판단해볼 것 같아요.
- public 메서드와 그 내부의 private 메서드 배치같은 것도 자신만의 기준을 만들어보면 좋을 것 같아요.
- 확실히 메서드명이 명확해졌지만, 더욱 분리하고 명확하게 나눈다면, 수정할 때 매번 모든 메서드를 찾아다니며 어디를 바꿔야 할지 찾는 것이 아니라 `어떤 부분을 담당하는 곳이 문제일 것`이라고 떠오르고 빠르게 해결할 수 있을 것 같아요. 이때 단위 테스트도 함께 작성되어 있다면 더욱 빨리 찾을 수 있을 것 같아요.

> 우선은 여러 클래스에서 사용해도 무방한 유틸 메서드는 별도의 유틸 클래스에 배치했습니다.
다만, 추후에 여러 클래스에서 한 유틸 클래스에 접근하게 되어, 해당 유틸 클래스를 수정할 시 영향을 미치는 범위가 매우 커질 수 있는 것이 우려스럽습니다.

- 우선 `CalculatorUtil`에는 모두 정적 메서드만 존재하는 것 같은데 public 생성자를 직접 명시하신 이유가 있나요?
- 매번 동일한 작업을 위해 중복된 코드를 작성하면 불일치 문제가 발생할 수 있습니다. 이를 해결하기 위해 중복된 코드를 하나로 추출하면 불일치 문제는 줄어들지만, 그 코드가 여러 곳에 영향을 미칠 가능성도 생깁니다. 따라서 `단순히 "여러 곳에서 사용되니까 분리하자!"` 라는 이유만으로 공통화하기보다는, `정말로 추출할 가치가 있는지` 더 신중하게 판단하는 것이 중요하다고 생각합니다. 이 코드가 실제로 공통된 역할을 하는가?, 분리했을 때 유지보수가 더 편리해지는가? 등의 기준을 고민해보면 좋을 것 같습니다.

> 1. 문서화
> 
> 3단계 요구사항을 수행하는 메서드를 구현하는데, 메서드의 이름을 잘 짓는 것 만으로는 클라이언트 측에서 해당 메서드를 이해하기 힘들 것이라 생각했습니다.

- 우선 다양한 사람들이 제시하는 좋은 메서드명을 짓기 위한 방법을 많이 찾아볼 수 있어요. 우선 여러 방법을 경험해보면 좋을 것 같습니다! 
- 아무래도 세상에는 없던 기능을 메서드명으로 표현하려고 하니 어려울 것 같아요. 작성해주신 주석도 충분히 도움이 될 것 같습니다!

> 2. 정규 표현식 처리
>
> 문제의 요구사항에 정확히 명시되어 있지 않지만, .과 $같이 정규 표현식에서 메타문자로 사용하는 값 역시 커스텀 구분자로 사용할 수 있어야 한다고 판단했습니다.
현재는 커스텀 구분자로 메타문자를 사용할 경우, 해당 메타문자 앞에 이스케이프를 붙여 일반 문자처럼 사용되도록 했습니다.
수정한 로직 역시 과제의 요구사항에 부적합한지 궁금합니다.

- 과제 요구사항에는 관련 내용에 대한 멘트가 없어서 큰 문제가 없을 것 같아요. 😄
- 이번처럼 더욱 다양한 예외 상황에 대해서 고민해보시면 더 견고한 코드가 될 것 같습니다!

> 3. 로깅
>
> 정규 표현식 관련 로직을 구현하는데 많은 시행착오를 겪어, 이를 빠르게 파악하고자 로그를 추가했습니다.
평소에는 스프링 프로젝트에서 Slf4J 라이브러리의 로거를 사용했는데, 해당 과제에서는 Java가 기본적으로 제공해주는 로거를 사용하고자 java.util.logging.Logger를 사용했습니다.
제가 작성한 로그 코드에 개선점이 있을지 궁금합니다.(불필요하게 사용하고 있거나, 알아보기 어렵거나 등등)

- 잘 사용하신 것 같습니다! 👍 

### 기타

- 마지막으로, 잘못된 질문은 없습니다. 질문을 통해 더 깊이 있는 학습을 할 수 있도록 제안드린 것뿐이니, 전혀 죄송해하실 필요 없어요! 오히려 이렇게 고민하고 질문하는 과정이 정말 중요하다고 생각해요. 이번 미션도 고생하셨습니다. 마감 기한이기 때문에 우선 머지요청 드리겠습니다 편하게 답변 남겨주셔도 됩니다! 😄

YehyeokBang reviewed (commented):
`Calculator` 객체가 많이 무거운 것 같은데, 일반적인 계산기와 문자열 계산기를 분리해보는 것은 어떨까요?

kjoon418 commented:
처음에는 사칙연산과 문자열 계산 둘 다 근본적으로 숫자 데이터를 받아 연산을 수행하는 것이므로, 계산기라는 한 클래스 안에 묶는 것이 바람직하다 생각했습니다.
그러나, 문자열 계산 기능으로 인해 클래스가 기존 계획보다 많이 무거워졌고, 두 기능 간에 공유하는 메서드도 거의 없으므로 각자의 클래스로 분리하는 것이 좋아 보입니다.
조언해 주셔서 정말 감사합니다!
